File: core/src/main/java/com/starshootercity/OrbOfOrigin.java
package com.starshootercity;

import com.starshootercity.events.PlayerSwapOriginEvent;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextDecoration;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.PrepareItemCraftEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.EquipmentSlot;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.Recipe;
import org.bukkit.inventory.ShapedRecipe;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.List;

public class OrbOfOrigin implements Listener {
    public static NamespacedKey orbKey = new NamespacedKey(OriginsReborn.getInstance(), "orb-of-origin");

    public static final ItemStack orb = new ItemStack(Material.NAUTILUS_SHELL) {{
        ItemMeta meta = getItemMeta();
        meta.getPersistentDataContainer().set(orbKey, OriginSwapper.BooleanPDT.BOOLEAN, true);
        meta.setCustomModelData(1);
        meta.displayName(
                Component.text(AddonLoader.getTextFor("item.origins.orb_of_origin", "Orb of Origin"))
                        .color(NamedTextColor.AQUA)
                        .decoration(TextDecoration.ITALIC, false)
        );
        setItemMeta(meta);
    }};

    @EventHandler
    @SuppressWarnings("ConstantConditions") // itemStack is null on some versions
    public void onPrepareItemCraft(PrepareItemCraftEvent event) {
        Recipe recipe = event.getRecipe();
        if (recipe != null) {
            if (recipe.getResult().getType() == Material.CONDUIT) {
                for (ItemStack itemStack : event.getInventory().getMatrix()) {
                    if (itemStack != null && itemStack.getItemMeta() != null) {
                        if (itemStack.getItemMeta().getPersistentDataContainer().has(orbKey, OriginSwapper.BooleanPDT.BOOLEAN)) {
                            event.getInventory().setResult(null);
                        }
                    }
                }
            }
        }
    }

    @SuppressWarnings("unchecked")
    public OrbOfOrigin() {
        Bukkit.removeRecipe(orbKey);
        FileConfiguration config = OriginsReborn.getInstance().getConfig();
        if (config.getBoolean("orb-of-origin.enable-recipe")) {
            ShapedRecipe shapedRecipe = new ShapedRecipe(orbKey, orb);
            shapedRecipe.shape("012", "345", "678");
            int i = 0;
            List<?> recipeData = config.getList("orb-of-origin.recipe");
            if (recipeData == null) return;
            for (Object line : recipeData) {
                for (String name : (List<String>) line) {
                    Material material = Material.matchMaterial(name);
                    if (material == null) material = Material.AIR;
                    if (material == Material.AIR) {
                        i++;
                        continue;
                    }
                    shapedRecipe.setIngredient(String.valueOf(i).charAt(0), material);
                    i++;
                }
            }
            Bukkit.addRecipe(shapedRecipe);
        }
    }

    @EventHandler
    public void onPlayerInteract(PlayerInteractEvent event) {
        if (event.getClickedBlock() != null) return;
        ItemStack item = event.getItem();
        if (item == null) return;
        ItemMeta meta = item.getItemMeta();
        if (meta != null) {
            if (meta.getPersistentDataContainer().has(orbKey, OriginSwapper.BooleanPDT.BOOLEAN)) {
                if (OriginSwapper.orbCooldown.containsKey(event.getPlayer())) {
                    if (System.currentTimeMillis() - OriginSwapper.orbCooldown.get(event.getPlayer()) < 500) {
                        return;
                    }
                }
                ItemMeta heldMeta = event.getPlayer().getInventory().getItemInMainHand().getItemMeta();
                EquipmentSlot hand = EquipmentSlot.OFF_HAND;
                if (heldMeta != null && heldMeta.getPersistentDataContainer().has(orbKey, OriginSwapper.BooleanPDT.BOOLEAN)) hand = EquipmentSlot.HAND;
                if (hand == EquipmentSlot.HAND) event.getPlayer().swingMainHand();
                else event.getPlayer().swingOffHand();
                if (OriginsReborn.getInstance().getConfig().getBoolean("orb-of-origin.random")) {
                    OriginSwapper.selectRandomOrigin(event.getPlayer(), PlayerSwapOriginEvent.SwapReason.ORB_OF_ORIGIN);
                    OriginSwapper.openOriginSwapper(event.getPlayer(), PlayerSwapOriginEvent.SwapReason.ORB_OF_ORIGIN, AddonLoader.origins.indexOf(OriginSwapper.getOrigin(event.getPlayer())), 0, false, true);
                    if (OriginsReborn.getInstance().getConfig().getBoolean("orb-of-origin.consume")) {
                        item.setAmount(item.getAmount() - 1);
                        event.getPlayer().getInventory().setItemInMainHand(item);
                    }
                } else {
                    Bukkit.getScheduler().scheduleSyncDelayedTask(OriginsReborn.getInstance(), () -> OriginSwapper.openOriginSwapper(event.getPlayer(), PlayerSwapOriginEvent.SwapReason.ORB_OF_ORIGIN, 0, 0));
                }
            }
        }
    }
}

File: core/src/main/java/com/starshootercity/OrbOfOrigin.kt
package com.starshootercity

import com.starshootercity.AddonLoader.getOrigins
import com.starshootercity.AddonLoader.getTextFor
import com.starshootercity.OriginsReborn.Companion.NMSInvoker
import com.starshootercity.OriginsReborn.Companion.instance
import com.starshootercity.events.PlayerSwapOriginEvent
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.PrepareItemCraftEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.EquipmentSlot
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.ShapedRecipe
import org.endera.enderalib.utils.async.ioDispatcher

class OrbOfOrigin : Listener {
    @EventHandler
    fun onInventoryClick(event: InventoryClickEvent) {
        if (event.currentItem == null || event.currentItem!!.itemMeta == null) return
        val meta = event.currentItem!!.itemMeta
        if (meta.persistentDataContainer
                .has(orbKey, OriginSwapper.BooleanPDT.BOOLEAN) && !meta.persistentDataContainer
                .has(updatedKey, OriginSwapper.BooleanPDT.BOOLEAN)
        ) {
            event.currentItem!!.setItemMeta(orb.itemMeta)
        }
    }

    @EventHandler  // itemStack is null on some versions
    fun onPrepareItemCraft(event: PrepareItemCraftEvent) {
        val recipe = event.recipe
        if (recipe != null) {
            if (recipe.result.type == Material.CONDUIT) {
                for (itemStack in event.inventory.matrix) {
                    if (itemStack.itemMeta != null) {
                        if (itemStack.itemMeta.persistentDataContainer
                                .has<Byte?, Boolean?>(orbKey, OriginSwapper.BooleanPDT.BOOLEAN)
                        ) {
                            event.inventory.result = null
                        }
                    }
                }
            }
        }
    }

    init {
        Bukkit.removeRecipe(orbKey)
        if (OriginsReborn.mainConfig.orbOfOrigin.enableRecipe) {
            val shapedRecipe = ShapedRecipe(orbKey, orb).apply {
                shape("012", "345", "678")
            }

            val recipeData: List<String> = OriginsReborn.mainConfig.orbOfOrigin.recipe.flatten()

            recipeData.forEachIndexed { index, materialName ->
                val material = Material.matchMaterial(materialName) ?: Material.AIR
                if (material != Material.AIR) {
                    shapedRecipe.setIngredient(index.toString()[0], material)
                }
            }
            Bukkit.addRecipe(shapedRecipe)
        }
    }


    @EventHandler
    fun onPlayerInteract(event: PlayerInteractEvent) {
        event.clickedBlock?.takeIf { it.type.isInteractable }?.let { return }

        if (!event.action.isRightClick) return

        val player = event.player
        val item = event.item ?: return
        val meta = item.itemMeta ?: return

        if (!meta.persistentDataContainer.has(orbKey, OriginSwapper.BooleanPDT.BOOLEAN)) return

        OriginSwapper.orbCooldown[player]?.let { lastUse ->
            if (System.currentTimeMillis() - lastUse < 500) return
        }

        val hand = if (player.inventory.itemInMainHand.itemMeta
                ?.persistentDataContainer
                ?.has(orbKey, OriginSwapper.BooleanPDT.BOOLEAN) == true
        ) EquipmentSlot.HAND else EquipmentSlot.OFF_HAND

        when (hand) {
            EquipmentSlot.HAND -> player.swingMainHand()
            else -> player.swingOffHand()
        }

        if (instance.config.getBoolean("orb-of-origin.consume")) {
            item.amount--
            player.inventory.setItemInMainHand(item)
        }

        var opened = false
        CoroutineScope(ioDispatcher).launch {
            for (layer in AddonLoader.layers) {
                OriginSwapper.setOrigin(
                    player,
                    null,
                    PlayerSwapOriginEvent.SwapReason.ORB_OF_ORIGIN,
                    false,
                    layer!!
                )
                if (opened) continue

                if (instance.config.getBoolean("orb-of-origin.random.$layer")) {
                    OriginSwapper.selectRandomOrigin(player, PlayerSwapOriginEvent.SwapReason.ORB_OF_ORIGIN, layer)
                    val origin = OriginSwapper.getOrigin(player, layer)
                    val index = getOrigins(layer).indexOf(origin)
                    OriginSwapper.openOriginSwapper(
                        player,
                        PlayerSwapOriginEvent.SwapReason.ORB_OF_ORIGIN,
                        index,
                        0,
                        false,
                        true,
                        layer
                    )
                } else {
                    OriginSwapper.openOriginSwapper(
                        player,
                        PlayerSwapOriginEvent.SwapReason.ORB_OF_ORIGIN,
                        0,
                        0,
                        layer
                    )
                }
                opened = true
            }
        }
    }


    companion object {
        @JvmField
        var orbKey: NamespacedKey = NamespacedKey(instance, "orb-of-origin")
        var updatedKey: NamespacedKey = NamespacedKey(instance, "updated-orb")

        @JvmField
        val orb: ItemStack = object : ItemStack(Material.NAUTILUS_SHELL) {
            init {
                var meta = itemMeta
                meta.persistentDataContainer.set<Byte?, Boolean?>(orbKey, OriginSwapper.BooleanPDT.BOOLEAN, true)
                meta.persistentDataContainer
                    .set<Byte?, Boolean?>(updatedKey, OriginSwapper.BooleanPDT.BOOLEAN, true)
                meta = NMSInvoker.setCustomModelData(meta, 1)
                meta.displayName(
                    Component.text(getTextFor("item.origins.orb_of_origin", "Orb of Origin"))
                        .color(NamedTextColor.AQUA)
                        .decoration(TextDecoration.ITALIC, false)
                )
                setItemMeta(meta)
            }
        }
    }
}
